### PR TITLE
Add open/closed range arguments for incremental

### LINKED ISSUE #1948: Support to customize incremental compare operators for sql_table
### Feature description

The sql_table source/resource is hard coded to use the operator ge (`>=`) when comparing data with incremental state (or le (`>=`) depending on the incremental function/sort order), we would like the option to configure it to use gt (`>`) / lt in stead.

### Are you a dlt user?

Yes, I run dlt in production.

### Use case

We have tables which are loaded seldom, but with large volumes, all with the same incremental key value. We would like to run dlt on top of these often, but not use resource re-loading the same data every time no new data has been added.

### Proposed solution

An option to the sql_table source/resource whether to use classic (e.g. operator.ge) or aggressive (operator.gt) incremental filtering on the extracted sql-table.

### Related issues

_No response_

### PR DESCRIPTION
**Pull request by steinitzu**
### Description
This allows configuring whether the incremental range is open or closed on both sides (start/end value).  
Translates to changing the operators between `> | >=` / `< | <=` in sql database source and in incremental filtering.   

Default is `start_range=closed` and `end_range=open`, meaning the exact initial value is included and the exact end value is excluded.    
With `start_range=open` you get `WHERE cursor > last_value` instead of `>=`.  
With `end_range=closed` you get `... cursor <= end_value` so non-overlapping chunks are still possible without the start_value deduplication logic.

### Related Issues

- Fixes #1948

### CODE REVIEW

File: dlt/extract/incremental/__init__.py [hunk 1]
```diff
@@ -111,6 +112,8 @@ class Incremental(ItemTransform[TDataItem], BaseConfiguration, Generic[TCursorVa
     row_order: Optional[TSortOrder] = None
     allow_external_schedulers: bool = False
     on_cursor_value_missing: OnCursorValueMissing = "raise"
+    range_start: TIncrementalRange = "closed"
```
Comment (rudolfix) [inline dlt/extract/incremental/__init__.py:115]: do not forget docstrings and docs

Comment (rudolfix) [reply dlt/extract/incremental/__init__.py:115]: also add this to `TIncrementalArgs` (so we can define those in REST API)

Comment (rudolfix) [reply dlt/extract/incremental/__init__.py:115]: also when start_range is open - disable boundary deduplication. there's no reason to deduplicate. there's no boundary overlap

File: dlt/extract/incremental/transform.py [hunk 2]
```diff
@@ -218,6 +227,8 @@ def __call__(
             ) from ex
         # new_value is "less" or equal to last_value (the actual max)
         if last_value == new_value:
+            if self.range_start == "open":
+                return None, False, False
```
Comment (rudolfix) [inline dlt/extract/incremental/transform.py:231]: right! maybe this disables deduplication as well...

File: dlt/extract/incremental/transform.py [hunk 3]
```diff
@@ -311,13 +322,19 @@ def __call__(
 
         if self.last_value_func is max:
             compute = pa.compute.max
-            end_compare = pa.compute.less
-            last_value_compare = pa.compute.greater_equal
+            end_compare = pa.compute.less if self.range_end == "open" else pa.compute.less_equal
```
Comment (rudolfix) [inline dlt/extract/incremental/transform.py:325]: we should cache those when creating object (?)

File: dlt/sources/sql_database/helpers.py [hunk 4]
```diff
@@ -87,11 +91,11 @@ def _make_query(self) -> SelectAny:
...
         if last_value_func is max:  # Query ordered and filtered according to last_value function
-            filter_op = operator.ge
-            filter_op_end = operator.lt
+            filter_op = operator.ge if self.range_start == "closed" else operator.gt
+            filter_op_end = operator.lt if self.range_end == "open" else operator.le
         elif last_value_func is min:
-            filter_op = operator.le
-            filter_op_end = operator.gt
+            filter_op = operator.le if self.range_start == "closed" else operator.lt
+            filter_op_end = operator.gt if self.range_end == "open" else operator.ge
```
Comment (rudolfix) [inline dlt/sources/sql_database/helpers.py:98]: we should document this in sql_database docs. we have separate chapter for incremental loading. 
for example if we load incrementally by id or high resolution timestamp or we do not expect stuff to be added (ie. if we have a cursor on day) it is better to keep range open (that disables deduplication and produces faster code)

**Review by rudolfix — CHANGES_REQUESTED** [review]
this looks pretty complete to me! let's work on docs and a few proposed improments

File: dlt/extract/incremental/transform.py (hunk 2 above)
Comment (steinitzu) [reply dlt/extract/incremental/transform.py:238]: Not completely. Added this check to the `deduplication_disabled` and a test for it, so it gets disabled now

File: dlt/extract/incremental/__init__.py (hunk 1 above)
Comment (steinitzu) [reply dlt/extract/incremental/__init__.py:125]: Added all here

File: dlt/sources/sql_database/helpers.py (hunk 4 above)
Comment (steinitzu) [reply dlt/sources/sql_database/helpers.py:121]: Wrote a section on the sql docs, hope it's clear.

**Review by rudolfix — APPROVED** [review]
this is very good now!

**Pull request merged**